### PR TITLE
Feature/app validation

### PR DESCRIPTION
**Pull request by emarkees**
## In this milestone i add the spec test file for the models
Screenshot (443)

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
+  belongs_to :user, class_name: 'User', foreign_key: 'user_id'
   has_many :comments
   has_many :likes
 
-  after_save :increment_posts_counter
+  validates :title, presence: true, length: { maximum: 250 }
 
-  def recent_comments(limit = 5)
-    comments.order(created_at: :desc).limit(limit)
+  def recent_comments
```
Comment (omar-muhamad) [inline app/models/post.rb]: - [x] As per these [requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=exceed%20250%20characters.-,CommentsCounter%20must%20be%20an%20integer%20greater%20than%20or%20equal%20to%20zero,LikesCounter%20must%20be%20an%20integer%20greater%20than%20or%20equal%20to%20zero.,-Add%20unit%20specs), please add validations here in the post model:
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.

File: spec/users.spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    it 'should validate_integer_is_positive(:posts_counter)' do
+      subject.posts_counter = -1
+      expect(subject).to_not be_valid
+    end
+
+    it 'expect posts_counter should be an integer' do
+      subject.posts_counter = 2.5
+      expect(subject).to_not be_valid
+    end
+  end
```
Comment (omar-muhamad) [inline spec/users.spec.rb]: - [x] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=equal%20to%20zero.-,Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.,-Need%20a%20big), Please add a unit test for the `most_recent_posts` method in user model.

File: spec/posts.spec.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    it 'should validate_integer(:comment_counter)' do
+      subject.comment_counter = 1.7
+      expect(subject).to be_valid
+    end
+
+    it 'should validate_integer_greater_than_zero(:comment_counter)' do
+      subject.comment_counter = 1
+      expect(subject).to be_valid
+    end
+  end
```
Comment (omar-muhamad) [inline spec/posts.spec.rb]: - [x] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=equal%20to%20zero.-,Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.,-Need%20a%20big), Please add a unit test for these 2 methods `recent_comments` and `update_post_counter` in post model.

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @emarkees,

### Nice job, you are almost there.

cat-driving-serious

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] The naming conventions in rails for test files is to be snake case, So please instead of making the test file like this `users.spec.rb` rename it to be like `users_spec.rb` to make the test cases run when you run the command `rspec spec` because now it's not working.

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/users.spec.rb (hunk 2 above)
Comment (emarkees) [reply spec/users.spec.rb]: @Omar-Muhamad done

File: app/models/post.rb (hunk 1 above)
Comment (emarkees) [reply app/models/post.rb]: Done

File: spec/posts.spec.rb (hunk 3 above)
Comment (emarkees) [reply spec/posts.spec.rb]: @Omar-Muhamad done

File: spec/posts.spec.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    it 'should validate_integer(:comment_counter)' do
+      subject.comment_counter = 1.7
+      expect(subject).to_not be_valid
+    end
+
+    it 'should validate_integer_greater_than_zero(:comment_counter)' do
+      subject.comment_counter = 1
+      expect(subject).to be_valid
+    end
+  end
```
Comment (KanzaTahreem) [inline spec/posts.spec.rb]: - [x] Good job for covering most of the previous reviewer's suggestions. However, I think you missed one from the review summary.  The following hasn't been implemented, please make sure to address this 👇
  
  image

- [x] Good job so far 👏👏However, I couldn't find the tests for the `recent_comments` and `update_post_counter` methods present in the `post` model [as suggested](https://github.com/emarkees/Gist_Blog/pull/3#discussion_r1306577214), Could you please take a look at this ? You can try something like below 👇


```ruby
context '#callbacks' do
  it 'should increment the user\'s posts_counter after save' do
    expect {
      Post.create(user_id: user.id, title: 'My first post', text: 'Here is my post', comment_counter: 0, like_counter: 0)
      user.reload
    }.to change(user, :posts_counter).by(1)
  end
end

context '#recent_comments' do
  it 'should return the most recent comments' do
    first_comment
    second_comment
    third_comment
    fourth_comment
    fifth_comment
    sixth_comment

    recent_comments = subject.recent_comments

    expect(recent_comments.length).to eq(5)
    expect(recent_comments).to eq([sixth_comment, fifth_comment, fourth_comment, third_comment, second_comment])
  end
end
```

**Review by KanzaTahreem — CHANGES_REQUESTED** [review]
# Changes Required 🔁

Hi @emarkees 

Great work on making the changes requested by a previous reviewer 👏! I really appreciate the effort you put in!  **_However, there are few this that still need to be fixed. But you are almost there!_** 💪 🏁

so close


## Highlights 🎯

There are some excellent points that you have made perfectly. Keep up the great work! 👏
- ✅ Good job with the validations
- ✅ Most of the suggestions are implemented
- ✅ Descriptive PR and README file


## Required Changes 🛑


_Please check the comments under the review._

## Optional suggestions ❕

N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👨‍💻💻

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/posts.spec.rb (hunk 4 above)
Comment (emarkees) [reply spec/posts.spec.rb]: @KanzaTahreem Thank you for the review. Changes made as suggested

Comment (KanzaTahreem) [reply spec/posts.spec.rb]: @emarkees I just sent you a message on Slack. Please take a look.

**Review by rotimiazeez — APPROVED** [review]
Hi @emarkees  👏🏽, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- Good job adding `Validations` to your `models` 👍🏽
- Good Job adding tests for all models 👌🏽
- All tests passed ✌🏽
- Good communication with the CR 😉
- No Linter error ✔️ 
- Good PR description 👍🏽
- Descriptive README 📖

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer, and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me(@rotimiazeez) in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561), you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by emarkees** [comment]
THanks @rotimiazeez @KanzaTahreem

**Pull request merged**